### DISCUSSION TITLE
urql nextjs ssg

### DISCUSSION THREAD

**Discussion by marceloverdijk** [urql-graphql/urql#2372] [Q&A] [answered]
What is the best way to use urql in a ssg next website?

I'm basically only interested to use urql in `getStaticPaths` and `getStatisProps` nowhere else, so e.g. no hooks.

I see many example using hooks like useQuery but can it used without hooks?

**Comment by JoviDeCroock** [comment, marked as answer]
You can just create a client and not care about the ssr-exchange if you don't want to 😅 just call client.query and inject it into props

```js
import { createClient, fetchExchange } from '@urql/core';

export const getStaticProps = () => {
  const client = createClient({ url: '', exchanges: [fetchExchange]})
  
  const data = await client.query().toPromise()
  
    // return data
}
```

**Reply by marceloverdijk** [reply]
Thx for feedback @JoviDeCroock ! Would that also work with typesafe queries? E.g. using https://www.graphql-code-generator.com/plugins/typescript-urql ?

**Reply by JoviDeCroock** [reply]
Yes, the query method accepts a generic so you can use YourGqlType as a generic

**Reply by kevinmamaqi** [reply]
Could you provide an example? I got a query generated by graphql-codegen named useGetMenuByIdQuery. How should I use it with client.query() fn?

**Reply by nandorojo** [reply]
For what it's worth, for my gql calls from the server, I circumvent urql altogether, instead using `typescript-graphql-request` codegen for these files.

For any queries that need to run inside of `getStaticProps` (or outside of React / the cache), I make the file name end in `.headless.graphql`. For example: `user.headless.graphql`.

Then, in my codegen file, I do this:

```ts
module.exports = {
  overwrite: true,
  schema: url,

  // ...other config

    './gql/index.ts': {
      documents: './**/*.graphql',
      plugins: [
        'typescript',
        'typescript-operations',
        'typescript-urql',
        'typescript-urql-graphcache',
      ], 
    },
    './gql/headless.ts': {
      documents: './**/*.headless.graphql',
      plugins: [
        'typescript',
        'typescript-operations',
        'typescript-graphql-request',
      ], 
    },
}
```

This results in 2 separate files created: one with all my URQL hooks, and another with my "headless" queries and mutations.

**Reply by kitten** [reply]
I'd still stick to `TypedDocumentNode`s here for new projects 👀 We may or may not be working on TypeScript tooling to make this experience a little more seamless and give ourselves more room to create an end-to-end framework experience. However, that'll very likely look more like generating typed GraphQL documents rather than full code .... Well, the migration path will be smoother.

If you're using the typed-document-node generator with GCG today however, you're basically bound to write a gql tag (or graphql file) then import the generated document with the appropriate types from the generated file and use that document.
Any part of urql that accepts a query document will then pick up types via generics 

(tl;dr: yes, the separate file thing bothers me immensely also, but we'll see how our alternative turns out)
